cmd/zrepl/mgmt_conn.c

```diff
@@ -538,6 +560,115 @@ uzfs_zvol_rebuild_status(uzfs_mgmt_conn_t *conn, zvol_io_hdr_t *hdrp,
 	return (reply_data(conn, &hdr, &status_ack, sizeof (status_ack)));
 }
 
+/*
+ * Iterate through all finished async tasks and send replies to clients.
+ */
+static int
+finish_async_tasks(void)
+{
+	uzfs_mgmt_conn_t *conn;
+	async_arg_t *async_task;
+	int rc;
+
+	mutex_enter(&async_tasks_mtx);
+	while ((async_task = SLIST_FIRST(&async_tasks)) != NULL) {
+		SLIST_REMOVE_HEAD(&async_tasks, task_next);
+		/* connection could have been removed in the meantime */
+		mutex_enter(&conn_list_mtx);
+		SLIST_FOREACH(conn, &uzfs_mgmt_conns, conn_next) {
```

Comment:
since async_task has conn info in ctx (if I am not wrong) then why we need to iterate over uzfs_mtm_conns ? May be I am missing something.

Reply:
because we don't know if the connection is still there when async task finishes (as is indicated by the comment above). We don't hold any locks or ref counts while executing async op which would guarantee conn to stay.